### PR TITLE
Migrate binding priority to int-based system

### PR DESCRIPTION
**Pull request by ZacSweers**
This allows more granular prioritization while also being ABI-compatible. See addition to the CHANGELOG for migration steps.

### CODE REVIEW

**Comment by ZacSweers** [comment]
I'm very confused by these test failures, `ContributesBindingGeneratorTest` all pass locally for me 🤔

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesBindingGeneratorTest.kt [hunk 1]
```diff
@@ -496,30 +534,30 @@ class ContributesBindingGeneratorTest(
...
       @ContributesBinding(Any::class, boundType = ParentInterface1::class) // Default case is NORMAL
-      @ContributesBinding(Any::class, boundType = ParentInterface2::class, priority = Priority.NORMAL)
-      @ContributesBinding(Unit::class, boundType = ParentInterface1::class, priority = Priority.HIGH)
-      @ContributesBinding(Unit::class, boundType = ParentInterface2::class, priority = Priority.HIGHEST)
+      @ContributesBinding(Any::class, boundType = ParentInterface2::class, priorityDeprecated = Priority.NORMAL)
+      @ContributesBinding(Unit::class, boundType = ParentInterface1::class, priorityDeprecated = Priority.HIGH)
+      @ContributesBinding(Unit::class, boundType = ParentInterface2::class, priorityDeprecated = Priority.HIGHEST)
       class ContributingInterface : ParentInterface1, ParentInterface2
       """,
       mode = mode,
```
Comment (RBusarow) [inline compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesBindingGeneratorTest.kt:542]:
```suggestion
      mode = mode,
      allWarningsAsErrors = false
```

Our default value for KCT's `-Werror` is tied to the `CI` environment variable.  So, that test's compilation is failing in CI with:

```
e: warnings found and -Werror specified
w: file:[...]/Source0.kt:4:58 'Priority' is deprecated. Use the new int-based priority
w: file:[...]/Source0.kt:10:91 'Priority' is deprecated. Use the new int-based priority
w: file:[...]/Source0.kt:11:92 'Priority' is deprecated. Use the new int-based priority
w: file:[...]/Source0.kt:12:92 'Priority' is deprecated. Use the new int-based priority
```

In hindsight, that `allWarningsAsErrors` probably shouldn't be toggled automagically, and we should probably have something in our test utils that asserts `exitCode == OK` by default.  I can follow up with those changes.

**Comment by ZacSweers** [comment]
Oh oops I committed before I saw you'd pushed new commits

**Comment by RBusarow** [comment]
😕  I did?  Didn't mean to!

**Review by RBusarow — APPROVED** [review]

**Pull request merged**